components/server/src/api/workspace-service-api.ts

```diff
@@ -92,4 +99,79 @@ export class WorkspaceServiceAPI implements ServiceImpl<typeof WorkspaceServiceI
+            throw new ApplicationError(ErrorCodes.UNIMPLEMENTED, "not implemented");
+        }
+        if (!req.organizationId || !uuidValidate(req.organizationId)) {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, "organizationId is required");
+        }
+        if (!req.editor) {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, "editor is required");
+        }
+        if (!req.source.value) {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, "source is required");
+        }
+        const contextUrl = req.source.value;
+        const user = await this.userService.findUserById(ctxUserId(), ctxUserId());
+        const { context, project } = await this.contextService.parseContext(user, contextUrl, {
+            projectId: req.configurationId,
+            organizationId: req.organizationId,
+            forceDefaultConfig: req.forceDefaultConfig,
+        });
+
+        const normalizedContextUrl = this.contextParser.normalizeContextURL(contextUrl);
+        const workspace = await this.workspaceService.createWorkspace(
+            {},
+            user,
+            req.organizationId,
+            project,
+            context,
+            normalizedContextUrl,
+        );
+
+        await this.workspaceService.startWorkspace({}, user, workspace.id, {
+            forceDefaultImage: req.forceDefaultConfig,
+            workspaceClass: req.workspaceClass,
+            ideSettings: {
+                defaultIde: req.editor.name,
+                useLatestVersion: req.editor.version === "latest",
+            },
+            clientRegionCode: ctxClientRegion(),
+        });
+
+        const info = await this.workspaceService.getWorkspace(ctxUserId(), workspace.id);
+        const response = new CreateAndStartWorkspaceResponse();
+        response.workspace = this.apiConverter.toWorkspace(info);
+        return response;
+    }
+
+    async startWorkspace(req: StartWorkspaceRequest): Promise<StartWorkspaceResponse> {
+        // We rely on FGA to do the permission checking
+        if (!req.workspaceId) {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, "workspaceId is required");
+        }
+        const user = await this.userService.findUserById(ctxUserId(), ctxUserId());
+        const { workspace, latestInstance: instance } = await this.workspaceService.getWorkspace(
+            ctxUserId(),
+            req.workspaceId,
+        );
+        if (instance && instance.status.phase !== "stopped") {
+            const info = await this.workspaceService.getWorkspace(ctxUserId(), workspace.id);
+            const response = new StartWorkspaceResponse();
+            response.workspace = this.apiConverter.toWorkspace(info);
+            return response;
+        }
+
+        await this.workspaceService.startWorkspace({}, user, workspace.id, {
+            forceDefaultImage: req.forceDefaultConfig,
+            clientRegionCode: ctxClientRegion(),
```

Comment:
I think we should region on proto file as well, and fallback to one from header if not provided for both create and start.

Reply:
Not sure what this value will be, since if we allow it then we need to tell values (or format of values) we expected

Reply:
We are going to show region on UI eventually. You can create an issue and add to `API hardening` milestone to consider.